File: research/e51_leg_parity.py
#!/usr/bin/env python3
"""E51 driver: emit one runtime-effective JIT string per arm, then run
research/e51_leg_parity.m over the scored affine-4/g64 shapes.

Arms patch the ONE line the campaign invariants pin, inside
`qmv_fast_crossrow_affine4_g64_wide`:

    r0      control, unchanged
    r0b     control with a comment-only change   NULL CONTROL: must stay equal
    d:<n>   R0's BF16 tree kept verbatim, its fp32 result scaled by 1 + 2^-n
            CALIBRATION: walks the comparator's minimum detectable effect
    r1      `(xm[0] + xm[1]) + (xm[2] + xm[3])`  pure reassociation
    r2      `xc[0] + xc[1] + xc[2] + xc[3]`      the fp32 values already in regs

`d:<n>` arms are calibration, never candidates: they widen no stored or
accumulated width and they are not proposals.

Research-only. Nothing here is on the scored path.

    research/e51_leg_parity.py --arms r0 r0b d:23 d:20 r1 r2
    research/e51_leg_parity.py --arms r0 r1 --shapes mlp.down --widths 7,8
"""

from __future__ import annotations

import argparse
import json
import pathlib
import shutil
import subprocess
import sys
import tempfile

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from jit_string_compile import assemble  # noqa: E402

ROOT = pathlib.Path(__file__).resolve().parent.parent
ENTRY_CELL = "affine_qmv_fast<bfloat16_t, 64, 4, false>"
HARNESS = ROOT / "research/e51_leg_parity.m"
R0_LINE = "sums[m] += xm[0] + xm[1] + xm[2] + xm[3];"


def arm_line(arm: str) -> str:
    if arm == "r0":
        return R0_LINE
    if arm == "r0b":
        return "/* e51 null control: comment only */ " + R0_LINE
    if arm == "r1":
        return "sums[m] += (xm[0] + xm[1]) + (xm[2] + xm[3]);"
    if arm == "r2":
        return "sums[m] += xc[0] + xc[1] + xc[2] + xc[3];"
    if arm.startswith("d:"):
        n = int(arm[2:])
        return ("sums[m] += static_cast<float>(xm[0] + xm[1] + xm[2] + xm[3]) "
                f"* (1.0f + 0x1p-{n}f);")
    raise SystemExit(f"unknown arm {arm!r}")


def emit_arm(arm: str, directory: pathlib.Path) -> pathlib.Path:
    source = assemble((ENTRY_CELL,), None)
    if source.count(R0_LINE) != 1:
        raise SystemExit(f"expected exactly one pinned bias line, found "
                         f"{source.count(R0_LINE)}")
    patched = source.replace(R0_LINE, arm_line(arm))
    path = directory / f"{arm.replace(':', '_')}.metal"
    path.write_text(patched)
    return path


def build_harness(directory: pathlib.Path) -> pathlib.Path:
    binary = directory / "e51_leg_parity"
    command = ["clang", "-fobjc-arc", "-O2", "-framework", "Metal",
               "-framework", "Foundation", "-o", str(binary), str(HARNESS)]
    subprocess.run(command, check=True)
    return binary


def summarise(payload: dict) -> None:
    print(f"\ndevice: {payload['device']}   repeats: {payload['repeats']}")
    arms = []
    for entry in payload["entries"]:
        if entry["arm"] not in arms:
            arms.append(entry["arm"])
    header = (f"{'arm':<8}{'cells':>7}{'legs equal':>12}{'legs diverge':>14}"
              f"{'max|d| legs':>14}{'max ulp':>9}{'wide!=ref':>11}"
              f"{'serial!=ref':>13}{'unstable':>10}")
    print(header)
    print("-" * len(header))
    for arm in arms:
        rows = [e for e in payload["entries"] if e["arm"] == arm]
        equal = sum(1 for e in rows if e["legs"]["equal"])
        diverge = len(rows) - equal
        max_abs = max(e["legs"]["max_abs_delta"] for e in rows)
        max_ulp = max(e["legs"]["max_ulp_delta"] for e in rows)
        wide_diff = sum(1 for e in rows if not e["wide_vs_ref"]["equal"])
        serial_diff = sum(1 for e in rows if not e["serial_vs_ref"]["equal"])
        unstable = sum(1 for e in rows if not e["self_stable"])
        print(f"{arm:<8}{len(rows):>7}{equal:>12}{diverge:>14}{max_abs:>14.4g}"
              f"{max_ulp:>9}{wide_diff:>11}{serial_diff:>13}{unstable:>10}")
    print("\nper-cell detail for arms that move something:")
    for entry in payload["entries"]:
        if entry["legs"]["equal"] and entry["wide_vs_ref"]["equal"]:
            continue
        legs, wide = entry["legs"], entry["wide_vs_ref"]
        print(f"  {entry['arm']:<8} {entry['shape']:<33} M={entry['m']}  "
              f"legs {'EQUAL' if legs['equal'] else 'DIVERGES'} "
              f"({legs['mismatches']}/{legs['compared']}, "
              f"frac {legs['mismatch_fraction']:.3g}, "
              f"max|d| {legs['max_abs_delta']:.3g}, ulp<={legs['max_ulp_delta']})  "
              f"wide_vs_ref {'equal' if wide['equal'] else 'differs'} "
              f"({wide['mismatches']}/{wide['compared']}, "
              f"frac {wide['mismatch_fraction']:.3g})")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--arms", nargs="+", default=["r0", "r0b", "r1", "r2"])
    parser.add_argument("--widths", default="3,4,5,6,7,8,9")
    parser.add_argument("--shapes")
    parser.add_argument("--repeats", type=int, default=2)
    parser.add_argument("--json", type=pathlib.Path,
                        default=ROOT / "research/e51-artifacts/leg-parity.json")
    args = parser.parse_args()

    if args.arms[0] != "r0":
        raise SystemExit("the first arm must be r0: it is the reference every "
                         "wide_vs_ref and serial_vs_ref comparison uses")

    args.json.parent.mkdir(parents=True, exist_ok=True)
    with tempfile.TemporaryDirectory(prefix="e51-legs-") as directory:
        work = pathlib.Path(directory)
        binary = build_harness(work)
        command = [str(binary), "--widths", args.widths,
                   "--repeats", str(args.repeats), "--json", str(args.json)]
        if args.shapes:
            command += ["--shapes", args.shapes]
        for arm in args.arms:
            command += ["--arm", f"{arm}={emit_arm(arm, work)}"]
        subprocess.run(command, check=True)
        shutil.rmtree(work / "__pycache__", ignore_errors=True)

    payload = json.loads(args.json.read_text())
    summarise(payload)
    print(f"\nartifact: {args.json}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: research/e51_leg_parity.m
// E51 step 0a: do the SERIAL leg (`qmv_fast_impl`, M == 1) and the MTP leg
// (`qmv_fast_crossrow_affine4_g64_wide`, M in 3..9) produce BIT-IDENTICAL
// affine-4/g64 output on the scored shapes, and does a dose to the pinned BF16
// bias tree break that identity?
//
// Both legs live in ONE compiled entry point, `affine_qmv_fast<T,64,4,false>`:
// the width is carried by `ntg.x`, so M == 1 falls through the switch to
// `qmv_fast_impl` and M in 3..9 reaches the wide cross-row cells. This harness
// therefore needs no model weights and no Swift build. It compiles each arm's
// runtime-effective JIT string with the scored MTLCompileOptions (language 4.0,
// fast math OFF, no include path -- device.cpp:631) and, per (shape, width),
// reports three independent comparisons:
//
//   legs      wide output at width M   vs  M == 1 output of the SAME arm
//   wide_r0   wide output at width M   vs  wide output of the reference arm
//   serial_r0 M == 1 output            vs  M == 1 output of the reference arm
//
// `legs` is the exactness question. `wide_r0` is the dose-bite check: it proves
// the arm changed the numbers the scored wide cell actually produces, which no
// AIR diff can establish. `serial_r0` must stay true for every arm, because the
// patch is confined to the wide cell; if it ever goes false the patch leaked.
//
// Untimed on purpose. Token streams and kernel outputs are deterministic, so no
// thermal gate applies and GPU contention cannot change a bit pattern.
//
// Research-only. Nothing here is on the scored path.
//
//   clang -fobjc-arc -O2 -framework Metal -framework Foundation \
//         -o /tmp/e51_leg_parity research/e51_leg_parity.m

#import <Foundation/Foundation.h>
#import <Metal/Metal.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static inline uint16_t f32_to_bf16(float f) {
  uint32_t u;
  memcpy(&u, &f, 4);
  u += 0x7fffu + ((u >> 16) & 1u);  // round-to-nearest-even
  return (uint16_t)(u >> 16);
}

static inline float bf16_to_f32(uint16_t h) {
  uint32_t u = (uint32_t)h << 16;
  float f;
  memcpy(&f, &u, 4);
  return f;
}

static inline uint32_t xorshift32(uint32_t *s) {
  uint32_t x = *s;
  x ^= x << 13;
  x ^= x >> 17;
  x ^= x << 5;
  *s = x;
  return x;
}

static inline float unit(uint32_t *s) {
  return (float)(xorshift32(s) >> 8) / (float)(1u << 24);
}

typedef struct {
  const char *name;
  int k;
  int n;
} Shape;

// The affine-4/g64 projection shapes of the scored Qwen 3.8 27B tree that reach
// the `out_vec_size >= 4096` wide branch (research/qmv_cost_curve.py).
static const Shape kShapes[] = {
    {"linear_attn.in_proj_fused_qkvzba", 5120, 16480},
    {"linear_attn.out_proj", 6144, 5120},
    {"full_attn.qkv_proj_fused", 5120, 14336},
    {"full_attn.o_proj", 6144, 5120},
    {"mlp.gate_up_fused", 5120, 34816},
    {"mlp.down", 17408, 5120},
    {"head.lm_head", 5120, 248320},
};
static const int kShapeCount = (int)(sizeof(kShapes) / sizeof(kShapes[0]));

typedef struct {
  id<MTLBuffer> w;
  id<MTLBuffer> scales;
  id<MTLBuffer> biases;
  id<MTLBuffer> x;
  id<MTLBuffer> yWide;
  id<MTLBuffer> ySerial;
  id<MTLBuffer> inVec;
  id<MTLBuffer> outVec;
  int k;
  int n;
  int maxM;
} Operands;

static id<MTLComputePipelineState> buildArm(id<MTLDevice> device, NSString *path,
                                            NSString *label) {
  NSError *err = nil;
  NSString *src = [NSString stringWithContentsOfFile:path
                                            encoding:NSUTF8StringEncoding
                                               error:&err];
  if (!src) {
    fprintf(stderr, "e51_leg_parity: cannot read arm %s at %s\n",
            label.UTF8String, path.UTF8String);
    exit(1);
  }
  MTLCompileOptions *opts = [MTLCompileOptions new];
  if (@available(macOS 26.0, *)) {
    opts.languageVersion = MTLLanguageVersion4_0;
  } else {
    opts.languageVersion = MTLLanguageVersion3_1;
  }
  [opts setFastMathEnabled:NO];
  id<MTLLibrary> lib = [device newLibraryWithSource:src options:opts error:&err];
  if (!lib) {
    fprintf(stderr, "e51_leg_parity: arm %s failed to compile: %s\n",
            label.UTF8String,
            err ? err.localizedDescription.UTF8String : "unknown");
    exit(1);
  }
  id<MTLFunction> fn =
      [lib newFunctionWithName:@"affine_qmv_fast_bfloat16_t_64_4_false"];
  if (!fn) {
    fprintf(stderr, "e51_leg_parity: arm %s has no affine_qmv_fast entry\n",
            label.UTF8String);
    exit(1);
  }
  id<MTLComputePipelineState> pso =
      [device newComputePipelineStateWithFunction:fn error:&err];
  if (!pso) {
    fprintf(stderr, "e51_leg_parity: arm %s pipeline failed: %s\n",
            label.UTF8String,
            err ? err.localizedDescription.UTF8String : "unknown");
    exit(1);
  }
  fprintf(stderr, "e51_leg_parity: arm %-10s compiled  src=%zu bytes\n",
          label.UTF8String, (size_t)src.length);
  return pso;
}

static Operands makeOperands(id<MTLDevice> device, Shape shape, int maxM) {
  Operands o;
  o.k = shape.k;
  o.n = shape.n;
  o.maxM = maxM;
  const size_t words = (size_t)shape.n * (size_t)shape.k / 8;
  const size_t groups = (size_t)shape.n * (size_t)shape.k / 64;

  o.w = [device newBufferWithLength:words * 4 options:MTLResourceStorageModeShared];
  o.scales = [device newBufferWithLength:groups * 2
                                 options:MTLResourceStorageModeShared];
  o.biases = [device newBufferWithLength:groups * 2
                                 options:MTLResourceStorageModeShared];
  o.x = [device newBufferWithLength:(size_t)maxM * shape.k * 2
                            options:MTLResourceStorageModeShared];
  o.yWide = [device newBufferWithLength:(size_t)maxM * shape.n * 2
                                options:MTLResourceStorageModeShared];
  o.ySerial = [device newBufferWithLength:(size_t)maxM * shape.n * 2
                                  options:MTLResourceStorageModeShared];
  o.inVec = [device newBufferWithLength:4 options:MTLResourceStorageModeShared];
  o.outVec = [device newBufferWithLength:4 options:MTLResourceStorageModeShared];
  if (!o.w || !o.scales || !o.biases || !o.x || !o.yWide || !o.ySerial) {
    fprintf(stderr, "e51_leg_parity: allocation failed for %s\n", shape.name);
    exit(1);
  }

  uint32_t seed = 0x1234567u;
  uint32_t *wp = (uint32_t *)o.w.contents;
  for (size_t i = 0; i < words; i++) {
    wp[i] = xorshift32(&seed);
  }
  uint16_t *sp = (uint16_t *)o.scales.contents;
  uint16_t *bp = (uint16_t *)o.biases.contents;
  for (size_t i = 0; i < groups; i++) {
    float s = 0.004f + 0.004f * unit(&seed);
    sp[i] = f32_to_bf16(s);
    bp[i] = f32_to_bf16(-7.5f * s);
  }
  uint16_t *xp = (uint16_t *)o.x.contents;
  for (size_t i = 0; i < (size_t)maxM * shape.k; i++) {
    xp[i] = f32_to_bf16(2.0f * unit(&seed) - 1.0f);
  }
  memset(o.yWide.contents, 0, o.yWide.length);
  memset(o.ySerial.contents, 0, o.ySerial.length);
  *(int *)o.inVec.contents = shape.k;
  *(int *)o.outVec.contents = shape.n;
  return o;
}

// One dispatch of the shared entry point. `ntgX` selects the leg: 1 reaches
// `qmv_fast_impl`, 3..9 reach the wide cross-row cells.
static void encode(id<MTLComputeCommandEncoder> enc,
                   id<MTLComputePipelineState> pso, Operands *o,
                   id<MTLBuffer> y, size_t xOffset, size_t yOffset, int ntgX) {
  [enc setComputePipelineState:pso];
  [enc setBuffer:o->w offset:0 atIndex:0];
  [enc setBuffer:o->scales offset:0 atIndex:1];
  [enc setBuffer:o->biases offset:0 atIndex:2];
  [enc setBuffer:o->x offset:xOffset atIndex:3];
  [enc setBuffer:y offset:yOffset atIndex:4];
  [enc setBuffer:o->inVec offset:0 atIndex:5];
  [enc setBuffer:o->outVec offset:0 atIndex:6];
  [enc dispatchThreadgroups:MTLSizeMake((NSUInteger)ntgX,
                                        (NSUInteger)(o->n / 8), 1)
      threadsPerThreadgroup:MTLSizeMake(32, 2, 1)];
}

static void runOnce(id<MTLCommandQueue> queue, id<MTLComputePipelineState> pso,
                    Operands *o, int m) {
  memset(o->yWide.contents, 0, o->yWide.length);
  memset(o->ySerial.contents, 0, o->ySerial.length);
  id<MTLCommandBuffer> cb = [queue commandBuffer];
  id<MTLComputeCommandEncoder> enc = [cb computeCommandEncoder];
  encode(enc, pso, o, o->yWide, 0, 0, m);
  for (int row = 0; row < m; row++) {
    encode(enc, pso, o, o->ySerial, (size_t)row * o->k * 2,
           (size_t)row * o->n * 2, 1);
  }
  [enc endEncoding];
  [cb commit];
  [cb waitUntilCompleted];
  if (cb.status != MTLCommandBufferStatusCompleted) {
    fprintf(stderr, "e51_leg_parity: command buffer failed (status %ld)\n",
            (long)cb.status);
    exit(1);
  }
}

typedef struct {
  size_t compared;
  size_t mismatches;
  double maxAbsDelta;
  double maxRelDelta;
  long maxUlpDelta;
  size_t firstMismatch;
} Compare;

static Compare compareBits(const uint16_t *a, const uint16_t *b, size_t count) {
  Compare c = {count, 0, 0.0, 0.0, 0, SIZE_MAX};
  for (size_t i = 0; i < count; i++) {
    if (a[i] == b[i]) {
      continue;
    }
    c.mismatches++;
    if (c.firstMismatch == SIZE_MAX) {
      c.firstMismatch = i;
    }
    double av = bf16_to_f32(a[i]);
    double bv = bf16_to_f32(b[i]);
    double delta = fabs(av - bv);
    if (delta > c.maxAbsDelta) {
      c.maxAbsDelta = delta;
    }
    double denom = fabs(bv) > 0.0 ? fabs(bv) : 1.0;
    if (delta / denom > c.maxRelDelta) {
      c.maxRelDelta = delta / denom;
    }
    // bf16 bit patterns of the same sign are ordered, so the raw difference is
    // an ulp count. Opposite signs are reported as the sentinel -1.
    long ulps = ((a[i] ^ b[i]) & 0x8000u)
                    ? -1
                    : labs((long)(a[i] & 0x7fffu) - (long)(b[i] & 0x7fffu));
    if (ulps > c.maxUlpDelta) {
      c.maxUlpDelta = ulps;
    }
  }
  return c;
}

static void printCompare(FILE *out, const char *key, Compare c) {
  fprintf(out,
          "\"%s\":{\"equal\":%s,\"compared\":%zu,\"mismatches\":%zu,"
          "\"mismatch_fraction\":%.6g,\"max_abs_delta\":%.9g,"
          "\"max_rel_delta\":%.6g,\"max_ulp_delta\":%ld,"
          "\"first_mismatch\":%lld}",
          key, c.mismatches == 0 ? "true" : "false", c.compared, c.mismatches,
          c.compared ? (double)c.mismatches / (double)c.compared : 0.0,
          c.maxAbsDelta, c.maxRelDelta, c.maxUlpDelta,
          c.firstMismatch == SIZE_MAX ? -1LL : (long long)c.firstMismatch);
}

int main(int argc, const char *argv[]) {
  @autoreleasepool {
    NSMutableArray<NSString *> *labels = [NSMutableArray array];
    NSMutableArray<NSString *> *paths = [NSMutableArray array];
    NSMutableArray<NSNumber *> *widths = [NSMutableArray array];
    NSMutableSet<NSString *> *shapeFilter = [NSMutableSet set];
    const char *jsonPath = NULL;
    int repeats = 1;

    for (int i = 1; i < argc; i++) {
      if (!strcmp(argv[i], "--arm") && i + 1 < argc) {
        NSString *spec = @(argv[++i]);
        NSRange eq = [spec rangeOfString:@"="];
        if (eq.location == NSNotFound) {
          fprintf(stderr, "e51_leg_parity: --arm wants label=path\n");
          return 2;
        }
        [labels addObject:[spec substringToIndex:eq.location]];
        [paths addObject:[spec substringFromIndex:eq.location + 1]];
      } else if (!strcmp(argv[i], "--widths") && i + 1 < argc) {
        for (NSString *w in [@(argv[++i]) componentsSeparatedByString:@","]) {
          [widths addObject:@(w.intValue)];
        }
      } else if (!strcmp(argv[i], "--shapes") && i + 1 < argc) {
        for (NSString *s in [@(argv[++i]) componentsSeparatedByString:@","]) {
          [shapeFilter addObject:s];
        }
      } else if (!strcmp(argv[i], "--repeats") && i + 1 < argc) {
        repeats = atoi(argv[++i]);
      } else if (!strcmp(argv[i], "--json") && i + 1 < argc) {
        jsonPath = argv[++i];
      } else {
        fprintf(stderr, "e51_leg_parity: unknown argument %s\n", argv[i]);
        return 2;
      }
    }
    if (labels.count == 0) {
      fprintf(stderr, "e51_leg_parity: at least one --arm label=path required\n");
      return 2;
    }
    if (widths.count == 0) {
      for (int m = 3; m <= 9; m++) {
        [widths addObject:@(m)];
      }
    }

    id<MTLDevice> device = MTLCreateSystemDefaultDevice();
    id<MTLCommandQueue> queue = [device newCommandQueue];
    fprintf(stderr, "e51_leg_parity: device %s\n", device.name.UTF8String);

    NSMutableArray<id<MTLComputePipelineState>> *psos = [NSMutableArray array];
    for (NSUInteger a = 0; a < labels.count; a++) {
      [psos addObject:buildArm(device, paths[a], labels[a])];
    }

    FILE *out = jsonPath ? fopen(jsonPath, "w") : stdout;
    fprintf(out, "{\"device\":\"%s\",\"repeats\":%d,\"entries\":[",
            device.name.UTF8String, repeats);
    bool firstEntry = true;

    int maxM = 1;
    for (NSNumber *w in widths) {
      maxM = w.intValue > maxM ? w.intValue : maxM;
    }

    for (int s = 0; s < kShapeCount; s++) {
      Shape shape = kShapes[s];
      if (shapeFilter.count && ![shapeFilter containsObject:@(shape.name)]) {
        continue;
      }
      Operands o = makeOperands(device, shape, maxM);
      // Reference-arm outputs, kept per width so every later arm compares
      // against the same bytes.
      NSMutableDictionary<NSNumber *, NSData *> *refWide =
          [NSMutableDictionary dictionary];
      NSMutableDictionary<NSNumber *, NSData *> *refSerial =
          [NSMutableDictionary dictionary];

      for (NSUInteger a = 0; a < labels.count; a++) {
        for (NSNumber *wn in widths) {
          int m = wn.intValue;
          NSData *wideFirst = nil;
          NSData *serialFirst = nil;
          bool selfStable = true;
          for (int rep = 0; rep < repeats; rep++) {
            runOnce(queue, psos[a], &o, m);
            NSData *wide = [NSData dataWithBytes:o.yWide.contents
                                          length:(size_t)m * shape.n * 2];
            NSData *serial = [NSData dataWithBytes:o.ySerial.contents
                                            length:(size_t)m * shape.n * 2];
            if (!wideFirst) {
              wideFirst = wide;
              serialFirst = serial;
            } else if (![wide isEqualToData:wideFirst] ||
                       ![serial isEqualToData:serialFirst]) {
              selfStable = false;
            }
          }
          const uint16_t *wide = (const uint16_t *)wideFirst.bytes;
          const uint16_t *serial = (const uint16_t *)serialFirst.bytes;
          size_t count = (size_t)m * shape.n;

          if (a == 0) {
            refWide[wn] = wideFirst;
            refSerial[wn] = serialFirst;
          }
          Compare legs = compareBits(wide, serial, count);
          Compare wideRef = compareBits(wide, (const uint16_t *)refWide[wn].bytes,
                                        count);
          Compare serialRef =
              compareBits(serial, (const uint16_t *)refSerial[wn].bytes, count);

          fprintf(out, "%s\n{\"arm\":\"%s\",\"shape\":\"%s\",\"k\":%d,\"n\":%d,"
                       "\"m\":%d,\"self_stable\":%s,",
                  firstEntry ? "" : ",", labels[a].UTF8String, shape.name,
                  shape.k, shape.n, m, selfStable ? "true" : "false");
          firstEntry = false;
          printCompare(out, "legs", legs);
          fprintf(out, ",");
          printCompare(out, "wide_vs_ref", wideRef);
          fprintf(out, ",");
          printCompare(out, "serial_vs_ref", serialRef);
          fprintf(out, "}");
          fflush(out);

          fprintf(stderr,
                  "  %-10s %-33s M=%d  legs %s (%zu/%zu, max|d|=%.3g, ulp<=%ld)"
                  "  wide_vs_ref %s  serial_vs_ref %s  self_stable %s\n",
                  labels[a].UTF8String, shape.name, m,
                  legs.mismatches == 0 ? "EQUAL   " : "DIVERGES",
                  legs.mismatches, legs.compared, legs.maxAbsDelta,
                  legs.maxUlpDelta,
                  wideRef.mismatches == 0 ? "equal" : "differs",
                  serialRef.mismatches == 0 ? "equal" : "differs",
                  selfStable ? "yes" : "NO");
        }
      }
    }
    fprintf(out, "\n]}\n");
    if (jsonPath) {
      fclose(out);
    }
  }
  return 0;
}
